lib/services/role.go

```diff
@@ -1307,6 +1311,52 @@ func (set RoleSet) CheckImpersonate(currentUser, impersonateUser types.User, imp
 	return trace.AccessDenied("access denied to '%s' to impersonate user '%s' and roles '%s'", currentUser.GetName(), impersonateUser.GetName(), roleNames(impersonateRoles))
 }
 
+// CheckImpersonateRoles validates that the current user can perform role-only
+// of the given roles. Role-only impersonation requires an allow rule with
+// roles but no users (and no user-less deny rules). All requested roles must
+// be allowed for the check to succeed.
+func (set RoleSet) CheckImpersonateRoles(currentUser types.User, impersonateRoles []types.Role) error {
+	ctx := &impersonateContext{
+		user: currentUser,
+	}
+	whereParser, err := newImpersonateWhereParser(ctx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// TODO: Unlike regular impersonation where all requested roles must be
+	// granted by a single impersonation role, it would be reasonable to
+	// request several roles whose `allow` conditions are split between
+	// several roles. Our initial use-case doesn't require this, so for now
+	// we'll assume all requested roles must be granted by a single `allow`.
+
+	// check deny: a single match on a deny rule prohibits access
+	for _, role := range set {
+		cond := role.GetImpersonateConditions(types.Deny)
+		matched, err := matchDenyRoleImpersonateCondition(cond, impersonateRoles)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matched {
+			return trace.AccessDenied("access denied to '%s' to impersonate roles '%s'", currentUser.GetName(), roleNames(impersonateRoles))
+		}
+	}
+
+	// check allow: if all match, allow impersonation
+	for _, role := range set {
+		cond := role.GetImpersonateConditions(types.Allow)
+		matched, err := matchAllowRoleImpersonateCondition(ctx, whereParser, cond, impersonateRoles)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+		if matched {
+			return nil
+		}
+	}
```

Comment:
Are the comment and the code contradicting here? If so which is right?

Reply:
This seems to allow impersonation if any of the roles match. Not all of them